tests/test_updater_top_level_update.py

```diff
@@ -0,0 +1,435 @@
+
+        # Hash mismatch error
+        with self.assertRaises(RepositoryError):
+            self._run_refresh()
+
+        self._assert_version_equals("timestamp", 3)
+        self._assert_version_equals("snapshot", 1)
+
+    def test_new_snapshot_unsigned(self) -> None:
+        # Check for an arbitrary software attack
+        self.sim.signers["snapshot"].clear()
+        with self.assertRaises(UnsignedMetadataError):
+            self._run_refresh()
+
+        self._assert_files_exist(["root", "timestamp"])
+
+    def test_new_snapshot_version_mismatch(self):
+        # Check against timestamp role’s snapshot version
+
+        # Increase snapshot version without updating timestamp
+        self.sim.snapshot.version += 1
+        with self.assertRaises(BadVersionNumberError):
+            self._run_refresh()
+
+    def test_new_snapshot_version_rollback(self) -> None:
+        # Check for a rollback attack
+        self.sim.snapshot.version = 2
+        self.sim.update_timestamp()
+        self._run_refresh()
+
+        self.sim.snapshot.version = 1
+        self.sim.update_timestamp()
+
+        with self.assertRaises(ReplayedMetadataError):
+            self._run_refresh()
+
+        self._assert_version_equals("snapshot", 2)
+
+    def test_new_snapshot_expired(self) -> None:
+        # Check for a freeze attack
+        self.sim.snapshot.expires = self.past_datetime
+        self.sim.update_snapshot()
+
+        with self.assertRaises(ExpiredMetadataError):
+            self._run_refresh()
+
+        self._assert_files_exist(["root", "timestamp"])
+
+    def test_new_targets_hash_mismatch(self) -> None:
+        # Check against snapshot role’s targets hashes
+
+        # Update snapshot with target's hashes
+        self.sim.compute_metafile_hashes_length = True
+        self.sim.update_snapshot()
+        self._run_refresh()
+
+        # Modify targets contents without updating
+        # snapshot's targets hashes
+        self.sim.targets.version += 1
+        self.sim.snapshot.meta[
+            "targets.json"
+        ].version = self.sim.targets.version
+        self.sim.snapshot.version += 1
+        self.sim.update_timestamp()
+
+        with self.assertRaises(RepositoryError):
+            self._run_refresh()
+
+        self._assert_version_equals("snapshot", 3)
+        self._assert_version_equals("targets", 1)
+
+    def test_new_targets_unsigned(self) -> None:
+        # Check for an arbitrary software attack
+        self.sim.signers["targets"].clear()
+        with self.assertRaises(UnsignedMetadataError):
+            self._run_refresh()
+
+        self._assert_files_exist(["root", "timestamp", "snapshot"])
+
+    def test_new_targets_version_mismatch(self):
+        # Check against snapshot role’s targets version
+
+        # Increase targets version without updating snapshot
+        self.sim.targets.version += 1
+        with self.assertRaises(BadVersionNumberError):
+            self._run_refresh()
```

Comment:
no _assert_files_exist() here?

Reply:
Oh yeah, I amended the last commit, thanks.